functions/src/modules/gmail.functions.ts

```diff
@@ -0,0 +1,273 @@
+import * as functions from 'firebase-functions';
+import * as admin from 'firebase-admin';
+import * as Google from 'googleapis';
+import * as querystring from 'querystring';
+import { Message } from 'firebase-functions/lib/providers/pubsub';
+
+const db = admin.database();
+
+export const oauth2init = functions.https.onRequest(
+  async (req: functions.Request, res: functions.Response) => {
+    const { client_key, secret, redirect } = functions.config().gmail;
+    const oauth2Client = new Google.google.auth.OAuth2(
+      client_key,
+      secret,
+      redirect
+    );
+    const authURL = oauth2Client.generateAuthUrl({
+      access_type: 'offline',
+      scope: ['https://www.googleapis.com/auth/gmail.modify'],
+      prompt: 'consent',
+    });
+    res.redirect(authURL);
+  }
+);
+
+const saveTokens = async (values: any) => {
+  const newGmailTokensRef = db.ref(`/gmail/${values.emailKey}`);
+  newGmailTokensRef.set({
+    oauth: {
+      token: values.accessToken,
+      refreshToken: values.refreshToken,
+    },
+    emailAddress: values.fullEmail,
+  });
+};
+
+export const oauth2callback = functions.https.onRequest(
+  async (req: functions.Request, res: functions.Response) => {
+    const { client_key, secret, redirect } = functions.config().gmail;
+    const oauth2Client = new Google.google.auth.OAuth2(
+      client_key,
+      secret,
+      redirect
+    );
+
+    const { tokens } = await oauth2Client.getToken(req.query.code);
+    oauth2Client.setCredentials(tokens);
+
+    const gm = await Google.google.gmail({
+      version: 'v1',
+      auth: oauth2Client,
+    });
+
+    const profile = await gm.users.getProfile({ userId: 'me' });
+    const emailAddressKey = profile.data.emailAddress
+      .split('@')[0]
+      .replace('.', '');
+
+    await saveTokens({
+      emailKey: emailAddressKey,
+      fullEmail: profile.data.emailAddress,
+      accessToken: tokens.access_token,
+      refreshToken: tokens.refresh_token,
+    });
+
+    res.redirect(
+      `/Gmail-initWatch?emailAddress=${querystring.escape(profile.data.emailAddress)}`
```

Comment:
I have re-considered this. Let's not redirect after authorization. Init function will be called separately. And there is no need to pass email, it should be taken from the config (card is updated). We want to monitor that specific mailbox, not any other.

Reply:
I've removed this for now.